cpp/src/search/contains_table.cu

```diff
@@ -0,0 +1,220 @@
+  accumulate_nullable_columns(table_view const& table)
+  {
+    for (auto const& col : table) {
+      accumulate(col);
+    }
+  }
+
+  auto release() { return std::move(result); }
+
+ private:
+  void accumulate(column_view const& col)
+  {
+    if (col.nullable()) { result.push_back(col); }
+    for (auto it = col.child_begin(); it != col.child_end(); ++it) {
+      if (it->size() == col.size()) { accumulate(*it); }
+    }
+  }
+};
+
+}  // namespace
+
+rmm::device_uvector<bool> contains(table_view const& haystack,
+                                   table_view const& needles,
+                                   null_equality compare_nulls,
+                                   nan_equality compare_nans,
+                                   rmm::cuda_stream_view stream,
+                                   rmm::mr::device_memory_resource* mr)
+{
+  // Use a hash map with key type is row hash values and map value type is `lhs_index_type` to store
+  // all row indices of the haystack table.
+  using static_multimap =
+    cuco::static_multimap<hash_value_type,
+                          lhs_index_type,
+                          cuda::thread_scope_device,
+                          rmm::mr::stream_allocator_adaptor<default_allocator<char>>,
+                          cuco::double_hashing<detail::DEFAULT_JOIN_CG_SIZE, hash_type, hash_type>>;
+
+  auto map = static_multimap(compute_hash_table_size(haystack.num_rows()),
+                             cuco::sentinel::empty_key{std::numeric_limits<hash_value_type>::max()},
+                             cuco::sentinel::empty_value{lhs_index_type{detail::JoinNoneValue}},
+                             stream.value(),
+                             detail::hash_table_allocator_type{default_allocator<char>{}, stream});
+
+  auto const haystack_has_nulls = has_nested_nulls(haystack);
+  auto const needles_has_nulls  = has_nested_nulls(needles);
+
+  // Insert all row hash values and indices of the haystack table.
+  {
+    auto const hasher   = cudf::experimental::row::hash::row_hasher(haystack, stream);
+    auto const d_hasher = hasher.device_hasher(nullate::DYNAMIC{haystack_has_nulls});
+
+    auto const kv_it = cudf::detail::make_counting_transform_iterator(
+      size_type{0},
+      make_pair_fn<lhs_index_type, decltype(d_hasher)>{d_hasher, map.get_empty_key_sentinel()});
+
+    // If the haystack table has nulls but they are compared unequal, don't insert them.
+    // Otherwise, it was known to cause performance issue:
+    // - https://github.com/rapidsai/cudf/pull/6943
+    // - https://github.com/rapidsai/cudf/pull/8277
+    if (haystack_has_nulls && compare_nulls == null_equality::UNEQUAL) {
+      // Gather all nullable columns at all levels from the right table.
+      auto const haystack_nullable_columns = accumulate_nullable_columns{haystack}.release();
+
+      [[maybe_unused]] auto const [row_bitmask, tmp] =
+        cudf::detail::bitmask_and(table_view{haystack_nullable_columns}, stream);
```

Comment:
```suggestion
      auto const row_bitmask =
        cudf::detail::bitmask_and(table_view{haystack_nullable_columns}, stream).first;
```

Reply:
This will make a deep copy of the bitmask.

Reply:
https://godbolt.org/z/KM688x471

`rmm::device_buffer` has move ctor ([here](https://github.com/rapidsai/rmm/blob/79f7e39ec89d4d9fdba46dd9f653cc7ccc523ef2/include/rmm/device_buffer.hpp#L168-L192)) so there should be no deep copies involved

Reply:
Oh yeah, so then we need to wrap that in `std::move()`.

Reply:
https://github.com/rapidsai/cudf/blob/f4f34287dadd9ee9a946b253aa045ba3256028b6/cpp/src/bitmask/null_mask.cu#L469

https://github.com/rapidsai/cudf/blob/4e66281f48c55735edb4b610e0f859ee2de32a75/cpp/include/cudf/detail/null_mask.cuh#L135

> so then we need to wrap that in std::move().

That's how `bitmask_and` implemented.